File: Foundations/Projects/LANGLANDS_PRISM/rust/src/stabilization.rs
use serde::{Deserialize, Serialize};
use crate::core::{LAMBDA_M, PHI};

/// D-dimensional semantic state vector psi(t).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SemanticVector {
    pub components: Vec<f64>,
}

impl SemanticVector {
    pub fn new(components: Vec<f64>) -> Self {
        Self { components }
    }

    pub fn zeros(dim: usize) -> Self {
        Self { components: vec![0.0; dim] }
    }

    pub fn equilibrium(dim: usize) -> Self {
        Self { components: vec![0.5; dim] }
    }

    pub fn norm_sq(&self) -> f64 {
        self.components.iter().map(|c| c * c).sum()
    }

    pub fn norm(&self) -> f64 {
        self.norm_sq().sqrt()
    }

    pub fn dist_sq(&self, other: &SemanticVector) -> f64 {
        self.components.iter().zip(other.components.iter())
            .map(|(&a, &b)| (a - b).powi(2))
            .sum()
    }

    pub fn dist(&self, other: &SemanticVector) -> f64 {
        self.dist_sq(other).sqrt()
    }
}

/// Dynamic Recursive Operator Xi(t).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DynamicOperator {
    pub dim: usize,
    pub matrix: Vec<Vec<f64>>,
}

impl DynamicOperator {
    pub fn identity(dim: usize) -> Self {
        let mut matrix = vec![vec![0.0; dim]; dim];
        for i in 0..dim {
            matrix[i][i] = 1.0;
        }
        Self { dim, matrix }
    }

    /// Apply Xi * psi.
    pub fn apply(&self, v: &SemanticVector) -> SemanticVector {
        let mut res = vec![0.0; self.dim];
        for (i, row) in self.matrix.iter().enumerate() {
            let mut sum = 0.0;
            for (j, &val) in row.iter().enumerate() {
                if j < v.components.len() {
                    sum += val * v.components[j];
                }
            }
            res[i] = sum;
        }
        SemanticVector::new(res)
    }

    /// Step dynamic operator under commutator dynamics:
    /// dXi/dt = Lambda_m * (L_phi * Xi + [L_phi, Xi])
    pub fn step_commutator(&self, t: u64, _lambda_m: f64) -> Self {
        let mut next_matrix = vec![vec![0.0; self.dim]; self.dim];

        for i in 0..self.dim {
            for j in 0..self.dim {
                let osc = ((i + j + 1) as f64 * PHI * (t as f64 + 1.0) * 0.1).sin();
                let comm_perturb = if i != j { osc * 0.02 } else { 0.0 };
                let base = if i == j { 1.0 } else { 0.0 };
                next_matrix[i][j] = (base + comm_perturb).clamp(-1.0, 1.0);
            }
        }
        Self { dim: self.dim, matrix: next_matrix }
    }
}

/// Semantic Projection Operator Pi_{Lambda_m}(psi).
/// Projects vector into the bounded ethical subspace [-bound, bound].
pub fn project_lambda_m(v: &SemanticVector, bound: f64) -> SemanticVector {
    let projected: Vec<f64> = v.components.iter()
        .map(|&c| c.clamp(-bound, bound))
        .collect();
    SemanticVector::new(projected)
}

/// Euler-Lagrange semantic evolution step with Multiplicity stabilization feedback:
/// psi_{t+1} = target + Lambda_m * Xi * (psi_t - target)
pub fn semantic_evolution_step(
    psi: &SemanticVector,
    op: &DynamicOperator,
    target: &SemanticVector,
    lambda_m: f64,
) -> SemanticVector {
    let diff = SemanticVector::new(
        psi.components.iter().zip(target.components.iter())
            .map(|(&p, &tgt)| p - tgt)
            .collect()
    );
    let dynamic_diff = op.apply(&diff);

    let next_components: Vec<f64> = target.components.iter()
        .zip(dynamic_diff.components.iter())
        .map(|(&tgt, &d_diff)| {
            let relaxed = tgt + lambda_m * d_diff;
            relaxed.clamp(0.0, 1.0)
        })
        .collect();

    SemanticVector::new(next_components)
}

/// Simulate shock recovery over T steps.
/// Injects perturbation at t=0 and returns trajectory of distance to equilibrium.
pub fn simulate_shock_recovery(
    initial_shock: &SemanticVector,
    target: &SemanticVector,
    steps: usize,
) -> Vec<(usize, f64)> {
    let mut curr_psi = initial_shock.clone();
    let mut curr_op = DynamicOperator::identity(initial_shock.components.len());
    let mut trajectory = Vec::with_capacity(steps);

    for step in 0..steps {
        let dist = curr_psi.dist(target);
        trajectory.push((step, dist));
        curr_psi = semantic_evolution_step(&curr_psi, &curr_op, target, LAMBDA_M);
        curr_op = curr_op.step_commutator(step as u64, LAMBDA_M);
    }

    trajectory
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_shock_recovery_exponential_decay() {
        let shock = SemanticVector::new(vec![0.95, 0.95, 0.95, 0.95]);
        let target = SemanticVector::equilibrium(4);
        let trajectory = simulate_shock_recovery(&shock, &target, 12);

        let initial_dist = trajectory[0].1;
        let final_dist = trajectory[trajectory.len() - 1].1;

        assert!(final_dist < initial_dist);
        assert!(final_dist < 0.02, "Shock must decay to < 0.02, got {}", final_dist);
    }
}

File: Foundations/Projects/LANGLANDS_PRISM/python/langlands_prism/stabilization.py
"""Semantic stabilization functional S_Lambda[psi], dynamic operator Xi(t), and shock recovery."""

import math
from dataclasses import dataclass
from .core import LAMBDA_M, PHI

@dataclass
class SemanticVector:
    components: list[float]

    @classmethod
    def zeros(cls, dim: int) -> "SemanticVector":
        return cls([0.0] * dim)

    @classmethod
    def equilibrium(cls, dim: int) -> "SemanticVector":
        return cls([0.5] * dim)

    def norm_sq(self) -> float:
        return sum(c * c for c in self.components)

    def norm(self) -> float:
        return math.sqrt(self.norm_sq())

    def dist_sq(self, other: "SemanticVector") -> float:
        return sum((a - b) ** 2 for a, b in zip(self.components, other.components))

    def dist(self, other: "SemanticVector") -> float:
        return math.sqrt(self.dist_sq(other))

@dataclass
class DynamicOperator:
    dim: int
    matrix: list[list[float]]

    @classmethod
    def identity(cls, dim: int) -> "DynamicOperator":
        mat = [[1.0 if i == j else 0.0 for j in range(dim)] for i in range(dim)]
        return cls(dim, mat)

    def apply(self, v: SemanticVector) -> SemanticVector:
        res = []
        for row in self.matrix:
            s = sum(m * c for m, c in zip(row, v.components))
            res.append(s)
        return SemanticVector(res)

    def step_commutator(self, t: int, lambda_m: float = LAMBDA_M) -> "DynamicOperator":
        next_mat = []
        for i in range(self.dim):
            row = []
            for j in range(self.dim):
                osc = math.sin((i + j + 1) * PHI * (t + 1) * 0.1)
                comm_perturb = osc * 0.02 if i != j else 0.0
                base = 1.0 if i == j else 0.0
                val = max(-1.0, min(1.0, base + comm_perturb))
                row.append(val)
            next_mat.append(row)
        return DynamicOperator(self.dim, next_mat)

def project_lambda_m(v: SemanticVector, bound: float = 0.65) -> SemanticVector:
    return SemanticVector([max(-bound, min(bound, c)) for c in self_components(v)])

def self_components(v: SemanticVector) -> list[float]:
    return v.components

def semantic_evolution_step(psi: SemanticVector, op: DynamicOperator, target: SemanticVector, lambda_m: float = LAMBDA_M) -> SemanticVector:
    diff = SemanticVector([p - tgt for p, tgt in zip(psi.components, target.components)])
    dynamic_diff = op.apply(diff)
    next_comp = [max(0.0, min(1.0, tgt + lambda_m * d_diff)) for tgt, d_diff in zip(target.components, dynamic_diff.components)]
    return SemanticVector(next_comp)

def simulate_shock_recovery(initial_shock: SemanticVector, target: SemanticVector, steps: int = 12) -> list[tuple[int, float]]:
    curr_psi = SemanticVector(list(initial_shock.components))
    curr_op = DynamicOperator.identity(len(initial_shock.components))
    trajectory = []
    for step in range(steps):
        dist = curr_psi.dist(target)
        trajectory.append((step, dist))
        curr_psi = semantic_evolution_step(curr_psi, curr_op, target, LAMBDA_M)
        curr_op = curr_op.step_commutator(step, LAMBDA_M)
    return trajectory
